File: apps/web/services/auth/oidcAdmin.ts
/**
 * Cliente REST da administração da configuração OIDC por organização
 * (feature 004). Implementação AGPL própria — NÃO reutiliza o fluxo SSO
 * Enterprise de `@services/auth/sso` (endpoints inexistentes na API OSS).
 *
 * Cobre os 4 endpoints de contracts/admin-oidc-api.md:
 *   GET/PUT/DELETE /orgs/{orgId}/oidc-config  e  POST /orgs/{orgId}/oidc-config/test
 */

import { getAPIUrl } from '@services/config/config'
import { getErrorMessage as getDetailMessage } from '@services/utils/ts/errorMessage'

// O segredo NUNCA aparece na leitura — apenas `secret_configured`.
export interface OIDCProviderConfigRead {
  id: number
  org_id: number
  issuer_url: string
  client_id: string
  secret_configured: boolean
  scopes: string
  enabled: boolean
  allowed_email_domains: string[]
  auto_provision_users: boolean
  default_role_id: number | null
  required_acr: string | null
  clock_skew_seconds: number
  created_by_user_id: number
  created_at: string | null
  updated_at: string | null
}

// PUT (upsert parcial). `client_secret` é write-only: omitir mantém o atual;
// string não vazia substitui; string vazia é rejeitada pela API (422).
export interface OIDCProviderConfigWrite {
  issuer_url?: string
  client_id?: string
  client_secret?: string
  scopes?: string
  enabled?: boolean
  allowed_email_domains?: string[]
  auto_provision_users?: boolean
  default_role_id?: number | null
  required_acr?: string | null
  clock_skew_seconds?: number
}

export interface OIDCConnectionTestResult {
  status: 'ok' | 'inacessivel' | 'invalida'
  detail: string
  discovered_endpoints: Record<string, string> | null
}

function authHeaders(accessToken: string): HeadersInit {
  return {
    Authorization: `Bearer ${accessToken}`,
    'Content-Type': 'application/json',
  }
}

/** Lê a configuração OIDC da org; `null` quando ainda não configurada (404). */
export async function getOIDCConfig(
  orgId: number,
  accessToken: string
): Promise<OIDCProviderConfigRead | null> {
  const response = await fetch(`${getAPIUrl()}orgs/${orgId}/oidc-config`, {
    method: 'GET',
    headers: authHeaders(accessToken),
    credentials: 'include',
  })
  if (response.status === 404) return null
  if (!response.ok) {
    const error = await response.json().catch(() => ({}))
    throw new Error(getDetailMessage(error?.detail, 'Falha ao carregar a configuração OIDC'))
  }
  return response.json()
}

/** Cria ou atualiza (upsert parcial) a configuração OIDC. */
export async function saveOIDCConfig(
  orgId: number,
  data: OIDCProviderConfigWrite,
  accessToken: string
): Promise<OIDCProviderConfigRead> {
  const response = await fetch(`${getAPIUrl()}orgs/${orgId}/oidc-config`, {
    method: 'PUT',
    headers: authHeaders(accessToken),
    credentials: 'include',
    body: JSON.stringify(data),
  })
  if (!response.ok) {
    const error = await response.json().catch(() => ({}))
    throw new Error(getDetailMessage(error?.detail, 'Falha ao salvar a configuração OIDC'))
  }
  return response.json()
}

/** Exclui a configuração (exige confirmação). Contas e vínculos são preservados. */
export async function deleteOIDCConfig(
  orgId: number,
  accessToken: string
): Promise<void> {
  const response = await fetch(
    `${getAPIUrl()}orgs/${orgId}/oidc-config?confirm=true`,
    {
      method: 'DELETE',
      headers: authHeaders(accessToken),
      credentials: 'include',
    }
  )
  if (!response.ok) {
    const error = await response.json().catch(() => ({}))
    throw new Error(getDetailMessage(error?.detail, 'Falha ao excluir a configuração OIDC'))
  }
}

/**
 * Testa a conexão com o provedor. `issuerUrl` opcional testa um endereço ainda
 * não salvo; omitido, testa o issuer da configuração salva.
 */
export async function testOIDCConnection(
  orgId: number,
  accessToken: string,
  issuerUrl?: string
): Promise<OIDCConnectionTestResult> {
  const response = await fetch(`${getAPIUrl()}orgs/${orgId}/oidc-config/test`, {
    method: 'POST',
    headers: authHeaders(accessToken),
    credentials: 'include',
    body: JSON.stringify(issuerUrl ? { issuer_url: issuerUrl } : {}),
  })
  if (!response.ok) {
    const error = await response.json().catch(() => ({}))
    throw new Error(getDetailMessage(error?.detail, 'Falha ao testar a conexão OIDC'))
  }
  return response.json()
}

File: apps/api/src/routers/oidc_admin.py
"""Endpoints administrativos da configuração OIDC por organização (feature 004).

Contrato em ``specs/004-admin-config-oidc/contracts/admin-oidc-api.md``:
GET/PUT/DELETE ``/orgs/{org_id}/oidc-config`` + POST
``/orgs/{org_id}/oidc-config/test``. Todo endpoint exige admin/mantenedor da
org (Princípio IV); respostas com ``Cache-Control: no-store``; o segredo do
cliente jamais aparece em resposta, log ou erro.
"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from pydantic import BaseModel
from sqlmodel.ext.asyncio.session import AsyncSession

from src.core.events.database import get_db_session
from src.db.oidc_provider_config import (
    OIDCConnectionTestResult,
    OIDCProviderConfigRead,
    OIDCProviderConfigWrite,
)
from src.db.users import PublicUser
from src.security.auth import get_current_user
from src.security.org_auth import require_org_admin
from src.services.auth.oidc_config import (
    delete_oidc_config,
    read_oidc_config,
    test_oidc_connection,
    upsert_oidc_config,
)
from src.services.security.rate_limiting import check_oidc_config_rate_limit

logger = logging.getLogger(__name__)

router = APIRouter()


class OIDCConnectionTestRequest(BaseModel):
    issuer_url: Optional[str] = None


def _no_store(response: Response) -> None:
    response.headers["Cache-Control"] = "no-store"


def _enforce_rate_limit(org_id: int, action: str) -> None:
    is_allowed, retry_after = check_oidc_config_rate_limit(org_id, action)
    if not is_allowed:
        minutos = max(1, retry_after // 60)
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail=(
                f"Muitas operações de configuração OIDC. Tente novamente em "
                f"cerca de {minutos} minuto{'s' if minutos != 1 else ''}."
            ),
            headers={"Retry-After": str(retry_after)},
        )


@router.get(
    "/{org_id}/oidc-config",
    response_model=OIDCProviderConfigRead,
    summary="Lê a configuração OIDC da organização",
    description=(
        "Retorna a configuração do provedor OIDC da organização. O segredo do "
        "cliente nunca é retornado — apenas `secret_configured`."
    ),
)
async def api_get_oidc_config(
    org_id: int,
    response: Response,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> OIDCProviderConfigRead:
    await require_org_admin(current_user.id, org_id, db_session)
    _no_store(response)
    config = await read_oidc_config(db_session, org_id)
    if config is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={
                "code": "OIDC_NAO_CONFIGURADO",
                "message": "Não há configuração OIDC para esta organização.",
            },
        )
    return config


@router.put(
    "/{org_id}/oidc-config",
    response_model=OIDCProviderConfigRead,
    summary="Cria ou atualiza a configuração OIDC (upsert parcial)",
    description=(
        "Upsert da configuração. Mudança de issuer passa por validação "
        "anti-SSRF e discovery antes de persistir. `client_secret` é "
        "write-only: ausente mantém o atual; não vazio substitui."
    ),
)
async def api_put_oidc_config(
    org_id: int,
    body: OIDCProviderConfigWrite,
    response: Response,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> OIDCProviderConfigRead:
    await require_org_admin(current_user.id, org_id, db_session)
    _enforce_rate_limit(org_id, "write")
    _no_store(response)
    return await upsert_oidc_config(db_session, org_id, body, current_user)


@router.delete(
    "/{org_id}/oidc-config",
    summary="Exclui a configuração OIDC da organização",
    description=(
        "Exige `confirm=true`. Remove apenas a configuração — contas, vínculos "
        "e identidades externas são preservados para eventual reativação."
    ),
)
async def api_delete_oidc_config(
    org_id: int,
    response: Response,
    confirm: bool = Query(False),
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> dict:
    await require_org_admin(current_user.id, org_id, db_session)
    _enforce_rate_limit(org_id, "delete")
    _no_store(response)
    if not confirm:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "code": "CONFIRMACAO_NECESSARIA",
                "message": (
                    "A exclusão da configuração exige confirmação explícita "
                    "(confirm=true). Contas e vínculos são preservados."
                ),
            },
        )
    removida = await delete_oidc_config(db_session, org_id, current_user)
    if not removida:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={
                "code": "OIDC_NAO_CONFIGURADO",
                "message": "Não há configuração OIDC para esta organização.",
            },
        )
    return {"deleted": True}


@router.post(
    "/{org_id}/oidc-config/test",
    response_model=OIDCConnectionTestResult,
    summary="Testa a conexão com o provedor OIDC",
    description=(
        "Executa o discovery contra o issuer informado (sem salvar) ou contra "
        "o issuer da configuração salva. Distingue provedor inacessível de "
        "configuração inválida."
    ),
)
async def api_test_oidc_connection(
    org_id: int,
    body: OIDCConnectionTestRequest,
    response: Response,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> OIDCConnectionTestResult:
    await require_org_admin(current_user.id, org_id, db_session)
    _enforce_rate_limit(org_id, "test")
    _no_store(response)
    return await test_oidc_connection(db_session, org_id, body.issuer_url)
